src/ints/xms.cpp

```diff
@@ -16,478 +17,843 @@
+	XMS_INVALID_DST_OFFSET  = 0xa6,
+	XMS_INVALID_LENGTH      = 0xa7,
+	XMS_INVALID_OVERLAP     = 0xa8, // TODO: add support for this error
+	XMS_PARITY_ERROR        = 0xa9,
+	XMS_BLOCK_NOT_LOCKED    = 0xaa,
+	XMS_BLOCK_LOCKED        = 0xab,
+	XMS_LOCK_COUNT_OVERFLOW = 0xac,
+	XMS_LOCK_FAILED         = 0xad, // TODO: when should this be reported?
+	UMB_ONLY_SMALLER_BLOCK  = 0xb0,
+	UMB_NO_BLOCKS_AVAILABLE = 0xb1,
+	UMB_INVALID_SEGMENT     = 0xb2, // TODO: when should this be reported?
+};
 
 struct XMS_Block {
-	Bitu	size;
-	MemHandle mem;
-	uint8_t	locked;
-	bool	free;
+	uint32_t size_kb     = 0;
+	MemHandle mem_handle = -1;
+	// locked blocks should not be quietly moved by the XMS driver
+	uint8_t lock_count = 0;
+	bool is_free       = true;
 };
 
 #ifdef _MSC_VER
 #pragma pack (1)
 #endif
-struct XMS_MemMove{
-	uint32_t length;
-	uint16_t src_handle;
+struct XMS_MemMove {
+	uint32_t length = 0;
+
+	uint16_t src_handle = 0;
 	union {
 		RealPt realpt;
 		uint32_t offset;
 	} src;
-	uint16_t dest_handle;
+	uint16_t dst_handle = 0;
 	union {
 		RealPt realpt;
 		uint32_t offset;
-	} dest;
+	} dst;
 
 } GCC_ATTRIBUTE(packed);
 #ifdef _MSC_VER
 #pragma pack ()
 #endif
 
+// ***************************************************************************
+// Variables
+// ***************************************************************************
+
+static struct {
+	bool enable_global = false;
+
+	uint32_t enable_local_count = 0;
+} a20;
+
+constexpr uint32_t A20MaxLocalCount = UINT32_MAX;
+
+static struct {
+	// TODO: HMA support for applications is not yet available in the core
+
+	bool is_available = false;
+
+	bool dos_has_control = true;
+	bool app_has_control = false;
+
+	uint16_t min_alloc_size = 0;
+} hma;
+
+static struct {
+	bool is_available = false;
+} umb;
+
+static struct {
+	bool is_available = false;
+
+	RealPt callback = 0;
 
-Bitu XMS_EnableA20(bool enable) {
-	uint8_t val = IO_Read	(0x92);
-	if (enable) IO_Write(0x92,val | 2);
-	else		IO_Write(0x92,val & ~2);
-	return 0;
+	XMS_Block handles[NumXmsHandles] = {};
+} xms;
+
+// ***************************************************************************
+// Generic helper routines
+// ***************************************************************************
+
+static uint32_t get_num_pages(const uint32_t size_kb)
+{
+	// Number of pages needed for the given memory size in KB
+	return size_kb / 4 + ((size_kb & 3) ? 1 : 0);
+}
+
+static uint32_t get_mem_free_total_kb()
+{
+	return static_cast<uint32_t>(MEM_FreeTotal() * 4);
+}
+
+static uint32_t get_mem_free_largest_kb()
+{
+	return static_cast<uint32_t>(MEM_FreeLargest() * 4);
 }
 
-Bitu XMS_GetEnabledA20(void) {
-	return (IO_Read(0x92)&2)>0;
+static uint32_t get_mem_highest_address()
+{
+	return static_cast<uint32_t>((MEM_TotalPages() * MemPageSize) - 1);
+}
+
+static void warn_umb_realloc()
+{
+	static bool first_time = true;
+	if (first_time) {
+		first_time = false;
+		LOG_WARNING("XMS: UMB realloc not implemented");
```

Comment:
Why not? 😄

Reply:
Short answer: I consider this out-of-scope of this PR :)

Long answer:
- might be non-trivial (look at UMB allocation code), I would have to investigate the whole `DOS_*MemAlloc*` routines (I'm not familiar with them)
- DOSBox-X does not implement it currently, so nowhere to steal the code from :)
- I have looked into JEMM386 source code to see how this is implemented - unfortunately, it only allows shrinking the block, without even returning no longer needed memory back to DOS
- probably rarely used - so far I didn't come across any software using it